lib/rex/post/postgresql/ui/console/command_dispatcher/client.rb

```diff
@@ -0,0 +1,168 @@
+# -*- coding: binary -*-
+
+require 'pathname'
+require 'reline'
+
+module Rex
+  module Post
+    module PostgreSQL
+      module Ui
+
+        ###
+        #
+        # Core PostgreSQL client commands
+        #
+        ###
+        class Console::CommandDispatcher::Client
+
+          include Rex::Post::PostgreSQL::Ui::Console::CommandDispatcher
+
+          #
+          # Initializes an instance of the core command set using the supplied console
+          # for interactivity.
+          #
+          # @param [Rex::Post::PostgreSQL::Ui::Console] console
+          def initialize(console)
+            super
+
+            @db_search_results = []
+          end
+
+          #
+          # List of supported commands.
+          #
+          def commands
+            cmds = {
+              'query'   => 'Run a raw SQL query',
+              'shell'   => 'Enter a raw shell where SQL queries can be executed',
+            }
+
+            reqs = {}
+
+            filter_commands(cmds, reqs)
+          end
+
+          def name
+            'PostgreSQL Client'
+          end
+
+          def help_args?(args)
+            return false unless args.instance_of?(::Array)
+
+            args.include?('-h') || args.include?('--help')
+          end
+
+          def cmd_shell_help
+            print_line 'Usage: shell'
+            print_line
+            print_line 'Go into a raw SQL shell where SQL queries can be executed.'
+            print_line 'To exit, type `exit`, `quit`, `end` or `stop`.'
+            print_line
+          end
+
+          def cmd_shell(*args)
+            if help_args?(args)
+              cmd_shell_help
+              return
+            end
+
+            stop_words = %w[stop s exit e end quit q].freeze
+
+            # Allow the user to query the DB in a loop.
+            finished = false
+            until finished
+              begin
+                # This needs to be here, otherwise the `ensure` block would reset it to the previous
+                # value after a single query, meaning future queries would have the default prompt_block.
+                prompt_proc_before = ::Reline.prompt_proc
+                ::Reline.prompt_proc = proc { |line_buffer| line_buffer.each_with_index.map { |_line, i| i > 0 ? 'SQL *> ' : 'SQL >> ' } }
+
+                # This will loop until it receives `true`.
+                raw_query = ::Reline.readmultiline('SQL >> ', use_history = true) do |multiline_input|
+                  # In the case only a stop word was input, exit out of the REPL shell
+                  finished = multiline_input.split.count == 1 && stop_words.include?(multiline_input.split.last)
+                  # Accept the input until the current line does not end with '\', similar to a shell
+                  finished || multiline_input.split.empty? || !multiline_input.split.last&.end_with?('\\')
```

Comment:
Not a blocker; This approach makes it hard to copy/pasta arbitrary SQL snippets into the prompt. We might need to implement a basic lexer/parser to make this UX better

i.e. psql prompt

```
psql postgres://postgres:password@127.0.0.1:9000
psql (14.10 (Homebrew))
Type "help" for help.

postgres=# CREATE TABLE cars (
  brand VARCHAR(255),
  model VARCHAR(255),
  year INT
);
CREATE TABLE
```

Versus our REPL:

```
msf6 auxiliary(scanner/postgres/postgres_login) > run postgres://postgres:password@127.0.0.1:9000/template1 createsession=true verbose=false

[!] No active DB -- Credential data will not be saved!
[+] 127.0.0.1:9000 - Login Successful: postgres:password@template1
[*] PostgreSQL session 4 opened (127.0.0.1:61231 -> 127.0.0.1:9000) at 2024-01-24 10:11:32 +0000
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf6 auxiliary(scanner/postgres/postgres_login) > sessions -i -1
[*] Starting interaction with 4...

PostgreSQL @ 127.0.0.1:9000 (template1) > shell
SQL >> CREATE TABLE cars (
[*] Running SQL Command: 'CREATE TABLE cars ('
[-] Query result: ERROR	VERROR	C42601	Msyntax error at end of input	P20	Fscan.l	L1168	Rscanner_yyerror

SQL >>   brand VARCHAR(255),
[*] Running SQL Command: 'brand VARCHAR(255),'
[-] Query result: ERROR	VERROR	C42601	Msyntax error at or near "brand"	P1	Fscan.l	L1176	Rscanner_yyerror

SQL >>   model VARCHAR(255),
[*] Running SQL Command: 'model VARCHAR(255),'
[-] Query result: ERROR	VERROR	C42601	Msyntax error at or near "model"	P1	Fscan.l	L1176	Rscanner_yyerror

SQL >>   year INT
[*] Running SQL Command: 'year INT'
[-] Query result: ERROR	VERROR	C42601	Msyntax error at or near "year"	P1	Fscan.l	L1176	Rscanner_yyerror

SQL >> );
[*] Running SQL Command: ');'
[-] Query result: ERROR	VERROR	C42601	Msyntax error at or near ")"	P1	Fscan.l	L1176	Rscanner_yyerror

SQL >> 
```